### ISSUE TITLE
Publish the GraphQL client in a separate package

### ISSUE THREAD

**Issue by francoisromain** [jaydenseric/graphql-react#48]
Hello Jayden

Thank you for this great module. I just switch from Apollo-client to graphQL-react and it feels much better!

I am using graphQl-react in a non-react app. Currently, I have to install react and react-dom to use it. 
Would you consider extracting the `GraphQL` class in a separate npm module with no dependencies on react?

Thank you

* * *

Also, here is an example code of the GraphQL client running in node. 
This could be helpful to have this example in the docs, 
at least for me :-). 

```js
const fetch = require('node-fetch')
const { GraphQL } = require('graphql-react')
const graphql = new GraphQL()

if (!globalThis.fetch) {
  globalThis.fetch = fetch
}

const apiFetch = query => async variables => {
  try {
    const res = await graphql.operate({
      operation: { query, variables },
      fetchOptionsOverride: options => {
        options.url = 'https://graphql-pokemon.now.sh'
      }
    })

    const value = await res.cacheValuePromise

    if (value.graphQLErrors) throw value.graphQLErrors[0]
    if (value.fetchError) throw value.fetchError
    if (value.httpError) throw value.httpError
    if (value.parseError) throw value.parseError

    return value.data
  } catch (e) {
    const errorMessage = `Erreur API : ${e.message}`
    console.error(e)
    throw errorMessage
  }
}

const pokemon = apiFetch(
  'query Pokemon($name: String!) { pokemon(name: $name) { image } }'
)

const run = async () => {
  const res = await pokemon({ name: 'pikachu' })

  console.log(res)
}

run()
```

**Comment by jaydenseric** [comment]
Thanks for the kind words :)

> Would you consider extracting the `GraphQL` class in a separate npm module with no dependencies on react?

Yes! That has been the plan for some time now; hopefully I'll be able to get around to it sometime soon. I also plan to add code examples for the [`GraphQL` instance method `operate`](https://github.com/jaydenseric/graphql-react#graphql-instance-method-operate).

Regarding usage in Node.js… While the `GraphQL` client has been designed from the start to work both in a server and browser environment, there really is not a huge benefit to using it in Node.js if it's not part of a SSR setup with cache hydration on the client, etc.

Personally I just make GraphQL `fetch` requests from scratch for server only scripts. While `GraphQL` is super lightweight, it is still doing work like checking for file uploads and populating cache that might be unnecessary for a server only script.

**Comment by francoisromain** [comment]
Thank you for your answer. 

That's a very good news that you plan to publish the graphql client as a separate library. 

Regarding the usage in node, I am in line with your point. In my specific case I use the graphql-client in a vuejs app (here: https://github.com/MTES-MCT/camino-ui/blob/master/src/api/_client.js). The example in nodejs is just a generic way to show  how to use it. There is probably a better way to do it.

**Comment by jaydenseric** [comment]
Closing because in [the new API that's coming](https://github.com/jaydenseric/graphql-react/issues/53), there is no longer a `GraphQL` client! React context will provide cache and loading stores, and management of them will be handled using very focused React hooks. This way only absolutely necessary code will be bundled for any given route.

Potentially there will be a utility function for making GraphQL `fetch` requests (it always resolves a cacheable value, even in case of client or server errors) that could be published, but we can consider doing that once the final API is ready.

**Comment by francoisromain** [comment]
Hello @jaydenseric 
Is it possible to use the v13 in a non-react app?

With v12, we were using `graphql.operate` (like in the first post of this thread). Is there a way to migrate to v13?

Thank you

**Comment by jaydenseric** [comment]
@francoisromain while you may still run into problems with the [`react` and `react-dom` peer dependencies](https://github.com/jaydenseric/graphql-react/blob/4ea092280531b4ef62adb02d365e4fbb23e418d8/package.json#L48-L51), there are standalone functions for making GraphQL `fetch` requests in the new API:

- [function `fetchGraphQL`](https://github.com/jaydenseric/graphql-react#function-fetchgraphql)
- [function `fetchOptionsGraphQL`](https://github.com/jaydenseric/graphql-react#function-fetchoptionsgraphql)

They will just get you a result in a nice cacheable format.

You could also directly use the cache related stuff:

- [class `Cache`](https://github.com/jaydenseric/graphql-react#class-cache)
- [function `cacheDelete`](https://github.com/jaydenseric/graphql-react#function-cachedelete)
- [function `cacheEntryDelete`](https://github.com/jaydenseric/graphql-react#function-cacheentrydelete)
- [function `cacheEntryPrune`](https://github.com/jaydenseric/graphql-react#function-cacheentryprune)
- [function `cacheEntrySet`](https://github.com/jaydenseric/graphql-react#function-cacheentryset)
- [function `cacheEntryStale`](https://github.com/jaydenseric/graphql-react#function-cacheentrystale)
- [function `cachePrune`](https://github.com/jaydenseric/graphql-react#function-cacheprune)
- [function `cacheStale`](https://github.com/jaydenseric/graphql-react#function-cachestale)

If you are really fancy, you could also directly use the loading system:

- [class `Loading`](https://github.com/jaydenseric/graphql-react#class-loading)
- [class `LoadingCacheValue`](https://github.com/jaydenseric/graphql-react#class-loadingcachevalue)

Basically everything except the React hooks and context could be used in a non React project.
